app/lib/foreman_deployments/planner_visitor.rb

```diff
@@ -0,0 +1,26 @@
+module ForemanDeployments
+  class PlannerVisitor < BaseDereferenceVisitor
+    def initialize(parent_action)
+      super()
+      @parent_action = parent_action
+    end
+
+    protected
+
+    def visit_task_definition(task)
+      super
+      task.plan(@parent_action)
```

Comment:
The output should be cached here too. Consider our example:
``` yaml
User1: 
  p1: v1
Usergroup1: 
  users: 
    - !reference
        task: User1
```
What will happen is the following:
1. `visit_task_definition` will be called for user1, the user task will be planned
2. since user doesn't have references, the code will return to `StackDefinition.accept` for the next task
3. `visit_task_definition` will be called for usergroup1
4. it has ref to user, so `get_dereference` will be called
5. `get_planned_output` will be called for user1
6. since the cache is empty, `task.plan` will be called again for the user.
...